File: mst prim.cpp
//program minimum spanning tree prim
#include<iostream>
#include<vector>
#include<unordered_set>
#include<queue>

using namespace std;

struct Edge
{
    int bobot;
    int node;
    int parent;

    Edge(int b ,int n, int p):bobot(b),node(n),parent(p){};

    bool operator>(const Edge& other) const {
        return bobot > other.bobot;
    }
};

void printMST(const vector<Edge>& MstEdges){
    int sum = 0;
    cout << "Edge pembentuk MST : ";
    for (const Edge& edge : MstEdges)
    {
        cout << edge.parent << " -> " << edge.node << "(bobotnya : " << edge.bobot << ")" << endl;
        sum+=edge.bobot;
    }
    cout << "total bobot MST : " << sum << endl;
}

vector<int> prim(int v,vector<vector<int>> adj[]){
    priority_queue<Edge, vector<Edge>, greater<Edge>> pq;//priority queue ato min heap untuk nyimpen bbt n edge
    vector<int> visited(v,0);//array visited

    pq.push(Edge(0, 0, -1));// inisialisasi
    vector<Edge> MstEdges;

    while (!pq.empty())// iterasi "tree"
    {
        Edge e = pq.top();
        pq.pop();
        int node = e.node;
        int weight = e.bobot;
        int ortu = e.parent;

        if (visited[node] == 1) continue;
        {
            visited[node] = 1;
            if (ortu!= -1)
            {
                MstEdges.push_back(Edge(weight, node, ortu));
            }
            

            for (auto it: adj[node])
            {
                int adjNode = it[0];
                int edW = it[1];
                if (!visited[adjNode])
                {
                    pq.push(Edge(edW,adjNode,node));// perjelas
                }
                
            }  
        } 
    }
    printMST(MstEdges);
}

int main() {
    int v = 7;
    vector<vector<int>> adj[v];
    adj[0] = {{1, 433}, {2, 473}, {3, 456}};
    adj[1] = {{0, 433}, {2, 447}, {4, 263}};
    adj[2] = {{0, 473}, {1, 447}, {3, 236}, {5, 305}};
    adj[3] = {{0, 456}, {2, 236}, {5, 440}, {6, 454}};
    adj[4] = {{1, 263}, {5, 555}};
    adj[5] = {{2, 305}, {3, 440}, {4, 555}, {6, 753}};
    adj[6] = {{3, 454}, {5, 753}};

    prim(v, adj);
    

    return 0;
}

File: discreet math II/java stuff/mstPrim.java
import java.util.*;

class Edge {
    int node;
    int bobot;
    int parent;

    public Edge(int bobot, int node, int parent) {
        this.node = node;
        this.bobot = bobot;
        this.parent = parent;
    }
}

public class mstPrim {
    static void printMST(ArrayList<Edge> MstEdges) {
        int sum = 0;
        System.out.println("MST prim : \n");
        for (Edge edge : MstEdges) {
            System.out.println(edge.parent + " -> " + edge.node + " (bobot: " + edge.bobot + ")");
            sum += edge.bobot;
        }
        System.out.println("Total sum: \n" + sum);
    }

    static void MST(int V, ArrayList<ArrayList<ArrayList<Integer>>> adj) {
        PriorityQueue<Edge> pq = new PriorityQueue<>((x, y) -> x.bobot - y.bobot);//priority queue utk simpan edge n bbt
        int[] vis = new int[V];// visited array
        ArrayList<Edge> mstEdges = new ArrayList<>();// array nyimpen edge pembuat mst

        pq.add(new Edge(0, 0, -1)); // initialization

        while (!pq.isEmpty()) { 
            Edge top = pq.poll();
            int node = top.node;
            int bbt = top.bobot;
            int ortu = top.parent;

            if (vis[node] == 1) continue;
            {
                if (ortu != -1) {
                    mstEdges.add(new Edge(bbt, node, ortu));
                }
                vis[node] = 1;
                for (ArrayList<Integer> edge : adj.get(node)) {
                    int adjNode = edge.get(0);
                    int edgeWeight = edge.get(1);
    
                    if (vis[adjNode] == 0) {
                        pq.add(new Edge(edgeWeight, adjNode, node));
                    }
                }
            }
        }
        printMST(mstEdges);
    }

    public static void main(String[] args) {
        int V = 7;
        ArrayList<ArrayList<ArrayList<Integer>>> adj = new ArrayList<>();

        adj.add(new ArrayList<>());
        adj.add(new ArrayList<>());
        adj.add(new ArrayList<>());
        adj.add(new ArrayList<>());
        adj.add(new ArrayList<>());
        adj.add(new ArrayList<>());
        adj.add(new ArrayList<>());

        adj.get(0).add(new ArrayList<>(List.of(1, 433)));
        adj.get(0).add(new ArrayList<>(List.of(2, 473)));
        adj.get(0).add(new ArrayList<>(List.of(3, 456)));
        adj.get(1).add(new ArrayList<>(List.of(2, 447)));
        adj.get(1).add(new ArrayList<>(List.of(4, 263)));
        adj.get(2).add(new ArrayList<>(List.of(3, 236)));
        adj.get(2).add(new ArrayList<>(List.of(5, 305)));
        adj.get(3).add(new ArrayList<>(List.of(5, 440)));
        adj.get(3).add(new ArrayList<>(List.of(6, 454)));
        adj.get(4).add(new ArrayList<>(List.of(5, 555)));
        adj.get(5).add(new ArrayList<>(List.of(6, 753)));



        MST(V, adj);
    }
}
